{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3daf57ba-bd56-47a1-b144-3744dbe521ec",
   "metadata": {},
   "source": [
    "# import_to_desktop\n",
    "Imports the spoke-genelab data and metadata csv files into a Neo4j Desktop database.\n",
    "\n",
    "Start the spoke-genelab Graph DBMS before you run this notebook!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "43efeca0-9798-4730-bcf4-ac44f9a7acbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "import pandas as pd\n",
    "from py2neo import Graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a165edce-3e85-4b8a-9dcc-8e45994833d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7568e807-6017-4b6a-991b-2fe72841324a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_http(url, filename):\n",
    "    if os.path.exists(filename):\n",
    "        os.remove(filename)\n",
    "    data = requests.get(url)\n",
    "    with open(filename, 'wb')as file:\n",
    "        file.write(data.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7baf0469-c454-4276-bee1-634da04b0bf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import neo4j_utils\n",
    "import neo4j_bulk_importer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee92b527-6931-4810-8b80-a474c4f09e02",
   "metadata": {},
   "source": [
    "### Import the Knowledge Graph\n",
    "CSV data and metadata files are uploaded into the Neo4j Graph database from the [kg](https://github.com/BaranziniLab/spoke_genelab/tree/main/kg) directory using the [kg-import](https://github.com/sbl-sdsc/kg-import) bulk upload scripts. For a description of the data organization and the specification of metadata [see](https://github.com/sbl-sdsc/kg-import/blob/main/README.md)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "df9643da-fedf-471d-ae1e-49b5d23b47fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "drop_database: '/Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/bin/cypher-shell' -d system -u neo4j -p neo4jdemo 'DROP DATABASE `spoke-genelab-v0.0.3` IF EXISTS;'\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8e7e0d9c583b40a4a1e38d2cab418ae4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Executing:   0%|          | 0/85 [00:00<?, ?cell/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run_bulk_import: cd '/Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import'; '/Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/bin/neo4j-admin' database import full spoke-genelab-v0.0.3 --overwrite-destination --skip-bad-relationships --skip-duplicate-nodes --multiline-fields --array-delimiter='|' @args.txt\n",
      "Neo4j version: 5.24.0\n",
      "Importing the contents of these files into /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/data/databases/spoke-genelab-v0.0.3:\n",
      "Nodes:\n",
      "  [Assay]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Assay_2025-20-04_n.csv\n",
      "\n",
      "  [Study]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Study_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Study_2025-20-04_n.csv\n",
      "\n",
      "  [Gene]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Gene_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Gene_2025-20-04_n.csv\n",
      "\n",
      "  [CellType]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_CellType_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/CellType_2025-20-04_n.csv\n",
      "\n",
      "  [MetaNode]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/MetaNode_n.csv\n",
      "\n",
      "  [Mission]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Mission_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Mission_2025-20-04_n.csv\n",
      "\n",
      "  [MGene]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_MGene_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/MGene_2025-20-04_n.csv\n",
      "\n",
      "  [Anatomy]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Anatomy_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Anatomy_2025-20-04_n.csv\n",
      "\n",
      "  [MethylationRegion]:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_MethylationRegion_n.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/MethylationRegion_2025-20-04_n.csv\n",
      "\n",
      "Relationships:\n",
      "  MEASURED_ASmMR:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-MEASURED_ASmMR-MethylationRegion_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Assay-MEASURED_ASmMR-MethylationRegion_2025-20-04_r.csv\n",
      "\n",
      "  PERFORMED_SpAS:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Study-PERFORMED_SpAS-Assay_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Study-PERFORMED_SpAS-Assay_2025-20-04_r.csv\n",
      "\n",
      "  CONDUCTED_MIcS:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Mission-CONDUCTED_MIcS-Study_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Mission-CONDUCTED_MIcS-Study_2025-20-04_r.csv\n",
      "\n",
      "  MetaRelationship:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/MetaRelationship_r.csv\n",
      "\n",
      "  INVESTIGATED_ASiCT:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-INVESTIGATED_ASiCT-CellType_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Assay-INVESTIGATED_ASiCT-CellType_2025-20-04_r.csv\n",
      "\n",
      "  METHYLATED_IN_MGmMR:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_MGene-METHYLATED_IN_MGmMR-MethylationRegion_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/MGene-METHYLATED_IN_MGmMR-MethylationRegion_2025-20-04_r.csv\n",
      "\n",
      "  INVESTIGATED_ASiA:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-INVESTIGATED_ASiA-Anatomy_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Assay-INVESTIGATED_ASiA-Anatomy_2025-20-04_r.csv\n",
      "\n",
      "  MEASURED_ASmMG:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-MEASURED_ASmMG-MGene_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/Assay-MEASURED_ASmMG-MGene_2025-20-04_r.csv\n",
      "\n",
      "  IS_ORTHOLOG_MGiG:\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_MGene-IS_ORTHOLOG_MGiG-Gene_r.csv\n",
      "  /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/MGene-IS_ORTHOLOG_MGiG-Gene_2025-20-04_r.csv\n",
      "\n",
      "\n",
      "Available resources:\n",
      "  Total machine memory: 16.00GiB\n",
      "  Free machine memory: 174.6MiB\n",
      "  Max heap memory : 910.5MiB\n",
      "  Max worker threads: 8\n",
      "  Configured max memory: 155.4MiB\n",
      "  High parallel IO: true\n",
      "\n",
      "Cypher type normalization is enabled (disable with --normalize-types=false):\n",
      "  Property type of 'start' normalized from 'int' --> 'long' in /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_MethylationRegion_n.csv\n",
      "  Property type of 'end' normalized from 'int' --> 'long' in /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_MethylationRegion_n.csv\n",
      "  Property type of 'dist_to_feature' normalized from 'int' --> 'long' in /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_MethylationRegion_n.csv\n",
      "  Property type of 'methylation_diff' normalized from 'float' --> 'double' in /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-MEASURED_ASmMR-MethylationRegion_r.csv\n",
      "  Property type of 'q_value' normalized from 'float' --> 'double' in /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-MEASURED_ASmMR-MethylationRegion_r.csv\n",
      "  Property type of 'log2fc' normalized from 'float' --> 'double' in /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-MEASURED_ASmMG-MGene_r.csv\n",
      "  Property type of 'adj_p_value' normalized from 'float' --> 'double' in /Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/header_Assay-MEASURED_ASmMG-MGene_r.csv\n",
      "Import starting\n",
      "  Page cache size: 153.5MiB\n",
      "  Number of worker threads: 8\n",
      "  Estimated number of nodes: 93.326\n",
      "  Estimated number of relationships: 26.020.225\n",
      "\n",
      "Importing nodes\n",
      ".......... .......... .......... .......... ..........   5% ∆1s 361ms [1s 361ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆56ms [1s 417ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆137ms [1s 555ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆188ms [1s 743ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆72ms [1s 816ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆130ms [1s 946ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆56ms [2s 2ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆514ms [2s 517ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆37ms [2s 554ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆184ms [2s 738ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆196ms [2s 935ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆22ms [2s 957ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆261ms [3s 218ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆99ms [3s 317ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆141ms [3s 458ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆145ms [3s 604ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆121ms [3s 726ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆137ms [3s 864ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆58ms [3s 922ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆339ms [4s 261ms]\n",
      "Imported 92,582 nodes in 4s 268ms\n",
      "Prepare ID mapper\n",
      ".......... .......... .......... .......... ..........   5% ∆187ms [187ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆7ms [195ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆7ms [202ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆7ms [209ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆6ms [216ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆7ms [223ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆137ms [360ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆2ms [363ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆1ms [365ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆1ms [366ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆1ms [367ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆72ms [439ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆0ms [440ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆2ms [442ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆0ms [443ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆4ms [447ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆3ms [450ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆2ms [453ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆2ms [456ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆2ms [458ms]\n",
      "Importing relationships\n",
      "Converting to intermediary format\n",
      ".......... .......... .......... .......... ..........   5% ∆2s 404ms [2s 404ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆1s 108ms [3s 513ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆1s 201ms [4s 714ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆1s 217ms [5s 932ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆1s 124ms [7s 56ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆1s 230ms [8s 286ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆1s 175ms [9s 462ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆1s 285ms [10s 747ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆2s 577ms [13s 325ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆1s 363ms [14s 689ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆1s 244ms [15s 933ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆1s 267ms [17s 201ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆1s 259ms [18s 461ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆1s 251ms [19s 713ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆1s 294ms [21s 7ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆1s 401ms [22s 408ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆1s 274ms [23s 683ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆1s 279ms [24s 963ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆1s 205ms [26s 169ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆1s 286ms [27s 455ms]\n",
      "  using configuration:Configuration[numberOfWorkers=8, temporaryPath=/Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/data/databases/spoke-genelab-v0.0.3/temp, numberOfTrackedDense=10000, applyBatchSize=64]\n",
      "Importing relationships\n",
      ".......... .......... .......... .......... ..........   5% ∆4s 610ms [4s 610ms]\n",
      ".......... .......... .......... .......... ..........  10% ∆2s 779ms [7s 390ms]\n",
      ".......... .......... .......... .......... ..........  15% ∆2s 544ms [9s 935ms]\n",
      ".......... .......... .......... .......... ..........  20% ∆1s 949ms [11s 884ms]\n",
      ".......... .......... .......... .......... ..........  25% ∆2s 248ms [14s 132ms]\n",
      ".......... .......... .......... .......... ..........  30% ∆2s 984ms [17s 117ms]\n",
      ".......... .......... .......... .......... ..........  35% ∆13s 888ms [31s 6ms]\n",
      ".......... .......... .......... .......... ..........  40% ∆1s 499ms [32s 505ms]\n",
      ".......... .......... .......... .......... ..........  45% ∆1s 725ms [34s 230ms]\n",
      ".......... .......... .......... .......... ..........  50% ∆6s 656ms [40s 887ms]\n",
      ".......... .......... .......... .......... ..........  55% ∆1s 699ms [42s 587ms]\n",
      ".......... .......... .......... .......... ..........  60% ∆8s 166ms [50s 753ms]\n",
      ".......... .......... .......... .......... ..........  65% ∆19s 779ms [1m 10s 533ms]\n",
      ".......... .......... .......... .......... ..........  70% ∆4s 376ms [1m 14s 910ms]\n",
      ".......... .......... .......... .......... ..........  75% ∆1s 869ms [1m 16s 779ms]\n",
      ".......... .......... .......... .......... ..........  80% ∆6s 567ms [1m 23s 346ms]\n",
      ".......... .......... .......... .......... ..........  85% ∆3s 922ms [1m 27s 269ms]\n",
      ".......... .......... .......... .......... ..........  90% ∆3s 207ms [1m 30s 476ms]\n",
      ".......... .......... .......... .......... ..........  95% ∆3s 786ms [1m 34s 263ms]\n",
      ".......... .......... .......... .......... .......... 100% ∆2s 265ms [1m 36s 528ms]\n",
      "Imported 51,895,799 relationships in 2m 5s 182ms\n",
      "Flushing stores\n",
      "Flush completed in 7s 547ms\n",
      "IMPORT DONE in 2m 19s 817ms.\n",
      "\n",
      "create_database: '/Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/bin/cypher-shell' -d system -u neo4j -p neo4jdemo 'CREATE OR REPLACE DATABASE `spoke-genelab-v0.0.3`;'\n",
      "\n",
      "add_indices: '/Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/bin/cypher-shell' -d spoke-genelab-v0.0.3 -u neo4j -p neo4jdemo -f '/Users/Peter/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-11a50ea0-7f7c-49f9-b49b-00ffe185cc16/import/indices.cypher'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "neo4j_bulk_importer.import_from_csv_to_neo4j_desktop(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11f36002-2b54-4aee-9928-aed61ccd30d9",
   "metadata": {},
   "source": [
    "### Connect to the local Neo4j Graph database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8c09e5cd-bdb7-49ba-87d9-264ee2da31dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "database = os.environ.get(\"NEO4J_DATABASE\")\n",
    "username = os.environ.get(\"NEO4J_USERNAME\")\n",
    "password = os.environ.get(\"NEO4J_PASSWORD\")\n",
    "stylesheet = os.environ.get(\"NEO4J_STYLESHEET\")\n",
    "\n",
    "graph = Graph(\"bolt://localhost:7687\", name=database, user=username, password=password)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02a25b4a-17ab-458a-af3b-3b4fcea68515",
   "metadata": {},
   "source": [
    "## Metadata <a class=\"anchor\" id=\"Metadata\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cb113bf-d9e4-436e-8f48-29045032c582",
   "metadata": {},
   "source": [
    "### Node metadata\n",
    "spoke_genelab KG is a self-describing KG. The MetaNodes and MetaRelationships define the structure of the KG and the properties of nodes and relationships. The query below lists the nodes and their properties."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9034786c-6a1f-4081-8d69-6cc1f594f610",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nodeName</th>\n",
       "      <th>identifier</th>\n",
       "      <th>material_name_1</th>\n",
       "      <th>technology</th>\n",
       "      <th>material_name_2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>factors_2</th>\n",
       "      <th>material_2</th>\n",
       "      <th>factors_1</th>\n",
       "      <th>material_1</th>\n",
       "      <th>...</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>in_intron</th>\n",
       "      <th>end</th>\n",
       "      <th>end_date</th>\n",
       "      <th>flight_program</th>\n",
       "      <th>space_program</th>\n",
       "      <th>start_date</th>\n",
       "      <th>project_type</th>\n",
       "      <th>project_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anatomy</td>\n",
       "      <td>UBERON Ontology ID (string)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Assay</td>\n",
       "      <td>GeneLab Data System GLDS-ID-MD5_hashcode(factors and materials) (string)</td>\n",
       "      <td>Prefered name from UEBERON for first material, e.g., cell type, tissue, organ (string)</td>\n",
       "      <td>Assay technology (string)</td>\n",
       "      <td>Prefered name from UEBERON for second material, e.g., cell type, tissue, organ (string)</td>\n",
       "      <td>Assay measurement type (string)</td>\n",
       "      <td>Second factor(s) in transcriptional profiling (string[])</td>\n",
       "      <td>Second material studied, e.g., cell type, tissue, organ (string)</td>\n",
       "      <td>First factor(s) in transcriptional profiling (string[])</td>\n",
       "      <td>First material studied, e.g., cell type, tissue, organ (string)</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CellType</td>\n",
       "      <td>Cell Ontology ID (string)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene ENTREZID (string)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGene</td>\n",
       "      <td>Gene ENTREZID (string)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MethylationRegion</td>\n",
       "      <td>Identifier for base pair range (string)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>Name or number of the chromosome (string)</td>\n",
       "      <td>Starting base pair position (int)</td>\n",
       "      <td>Indicates whether the region overlaps with an intron (boolean)</td>\n",
       "      <td>Ending base pair position (int)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Mission</td>\n",
       "      <td>Identifier for mission (string)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>End date of mission (date)</td>\n",
       "      <td>Type of flight program (string)</td>\n",
       "      <td>Sponsor of the mission (string)</td>\n",
       "      <td>Start date of mission (date)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Study</td>\n",
       "      <td>NASA Open Science Data Repository ID (string)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Type of project (string)</td>\n",
       "      <td>Title of project (string)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            nodeName  \\\n",
       "0            Anatomy   \n",
       "1              Assay   \n",
       "2           CellType   \n",
       "3               Gene   \n",
       "4              MGene   \n",
       "5  MethylationRegion   \n",
       "6            Mission   \n",
       "7              Study   \n",
       "\n",
       "                                                                 identifier  \\\n",
       "0                                               UBERON Ontology ID (string)   \n",
       "1  GeneLab Data System GLDS-ID-MD5_hashcode(factors and materials) (string)   \n",
       "2                                                 Cell Ontology ID (string)   \n",
       "3                                                    Gene ENTREZID (string)   \n",
       "4                                                    Gene ENTREZID (string)   \n",
       "5                                   Identifier for base pair range (string)   \n",
       "6                                           Identifier for mission (string)   \n",
       "7                             NASA Open Science Data Repository ID (string)   \n",
       "\n",
       "                                                                          material_name_1  \\\n",
       "0                                                                                           \n",
       "1  Prefered name from UEBERON for first material, e.g., cell type, tissue, organ (string)   \n",
       "2                                                                                           \n",
       "3                                                                                           \n",
       "4                                                                                           \n",
       "5                                                                                           \n",
       "6                                                                                           \n",
       "7                                                                                           \n",
       "\n",
       "                  technology  \\\n",
       "0                              \n",
       "1  Assay technology (string)   \n",
       "2                              \n",
       "3                              \n",
       "4                              \n",
       "5                              \n",
       "6                              \n",
       "7                              \n",
       "\n",
       "                                                                           material_name_2  \\\n",
       "0                                                                                            \n",
       "1  Prefered name from UEBERON for second material, e.g., cell type, tissue, organ (string)   \n",
       "2                                                                                            \n",
       "3                                                                                            \n",
       "4                                                                                            \n",
       "5                                                                                            \n",
       "6                                                                                            \n",
       "7                                                                                            \n",
       "\n",
       "                       measurement  \\\n",
       "0                                    \n",
       "1  Assay measurement type (string)   \n",
       "2                                    \n",
       "3                                    \n",
       "4                                    \n",
       "5                                    \n",
       "6                                    \n",
       "7                                    \n",
       "\n",
       "                                                  factors_2  \\\n",
       "0                                                             \n",
       "1  Second factor(s) in transcriptional profiling (string[])   \n",
       "2                                                             \n",
       "3                                                             \n",
       "4                                                             \n",
       "5                                                             \n",
       "6                                                             \n",
       "7                                                             \n",
       "\n",
       "                                                         material_2  \\\n",
       "0                                                                     \n",
       "1  Second material studied, e.g., cell type, tissue, organ (string)   \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "5                                                                     \n",
       "6                                                                     \n",
       "7                                                                     \n",
       "\n",
       "                                                 factors_1  \\\n",
       "0                                                            \n",
       "1  First factor(s) in transcriptional profiling (string[])   \n",
       "2                                                            \n",
       "3                                                            \n",
       "4                                                            \n",
       "5                                                            \n",
       "6                                                            \n",
       "7                                                            \n",
       "\n",
       "                                                        material_1  ...  \\\n",
       "0                                                                   ...   \n",
       "1  First material studied, e.g., cell type, tissue, organ (string)  ...   \n",
       "2                                                                   ...   \n",
       "3                                                                   ...   \n",
       "4                                                                   ...   \n",
       "5                                                                   ...   \n",
       "6                                                                   ...   \n",
       "7                                                                   ...   \n",
       "\n",
       "                                  chromosome  \\\n",
       "0                                              \n",
       "1                                              \n",
       "2                                              \n",
       "3                                              \n",
       "4                                              \n",
       "5  Name or number of the chromosome (string)   \n",
       "6                                              \n",
       "7                                              \n",
       "\n",
       "                               start  \\\n",
       "0                                      \n",
       "1                                      \n",
       "2                                      \n",
       "3                                      \n",
       "4                                      \n",
       "5  Starting base pair position (int)   \n",
       "6                                      \n",
       "7                                      \n",
       "\n",
       "                                                        in_intron  \\\n",
       "0                                                                   \n",
       "1                                                                   \n",
       "2                                                                   \n",
       "3                                                                   \n",
       "4                                                                   \n",
       "5  Indicates whether the region overlaps with an intron (boolean)   \n",
       "6                                                                   \n",
       "7                                                                   \n",
       "\n",
       "                               end                    end_date  \\\n",
       "0                                                                \n",
       "1                                                                \n",
       "2                                                                \n",
       "3                                                                \n",
       "4                                                                \n",
       "5  Ending base pair position (int)                               \n",
       "6                                   End date of mission (date)   \n",
       "7                                                                \n",
       "\n",
       "                    flight_program                    space_program  \\\n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "5                                                                     \n",
       "6  Type of flight program (string)  Sponsor of the mission (string)   \n",
       "7                                                                     \n",
       "\n",
       "                     start_date              project_type  \\\n",
       "0                                                           \n",
       "1                                                           \n",
       "2                                                           \n",
       "3                                                           \n",
       "4                                                           \n",
       "5                                                           \n",
       "6  Start date of mission (date)                             \n",
       "7                                Type of project (string)   \n",
       "\n",
       "               project_title  \n",
       "0                             \n",
       "1                             \n",
       "2                             \n",
       "3                             \n",
       "4                             \n",
       "5                             \n",
       "6                             \n",
       "7  Title of project (string)  \n",
       "\n",
       "[8 rows x 28 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (n:MetaNode) RETURN n;\n",
    "\"\"\"\n",
    "df = graph.run(query).to_data_frame()\n",
    "metadata = df[\"n\"].tolist()\n",
    "metadata = pd.DataFrame(metadata)\n",
    "metadata.fillna(\"\", inplace=True)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18414583-2ae9-44d3-92f6-f7ab4a24b258",
   "metadata": {},
   "source": [
    "### Metagraph <a class=\"anchor\" id=\"Metagraph\"></a>\n",
    "The metagraph shows the node labels and relationship types of the KG. Click on a node to display the node metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ee748598-4c5b-4542-b408-c870c121f9a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"\n",
    "MATCH p=(:MetaNode)-->(:MetaNode) RETURN p\n",
    "\"\"\"\n",
    "subgraph1 = graph.run(query).to_subgraph()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8f6ca2a8-cb5e-4b06-a382-91e194a2b491",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f9a4f828a560436c945f158788f62e08",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "CytoscapeWidget(cytoscape_layout={'name': 'cola', 'padding': 0, 'nodeSpacing': 65, 'nodeDimensionsIncludeLabel…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Rerun this cell if the layout doesn't look good\n",
    "widget1 = neo4j_utils.draw_graph(subgraph1, stylesheet)\n",
    "widget1.layout.height = \"1014px\"\n",
    "widget1.set_layout(name=\"cola\", padding=00, nodeSpacing=65, nodeDimensionsIncludeLabels=True, unconstrIter=15000)\n",
    "display(widget1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeaea2e5-4430-4d06-81e8-d8e492a6e918",
   "metadata": {},
   "source": [
    "### Number of Nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d505d4c2-f693-4fa3-a85b-e8c23c60d450",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of nodes: 92582\n"
     ]
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (n) RETURN COUNT(n);\n",
    "\"\"\"\n",
    "print(f\"Total number of nodes: {graph.evaluate(query)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9bec0c45-e6ee-43a6-886d-05ef32ccd7bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGene</td>\n",
       "      <td>57596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene</td>\n",
       "      <td>28269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assay</td>\n",
       "      <td>6178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MethylationRegion</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Study</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Anatomy</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Mission</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MetaNode</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CellType</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Node  Count\n",
       "0              MGene  57596\n",
       "1               Gene  28269\n",
       "2              Assay   6178\n",
       "3  MethylationRegion    344\n",
       "4              Study    125\n",
       "5            Anatomy     33\n",
       "6            Mission     22\n",
       "7           MetaNode      8\n",
       "8           CellType      7"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (n) RETURN labels(n)[0] AS Node, COUNT(n) AS Count\n",
    "ORDER BY Count DESC;\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b04b239b-b337-4897-967c-c6be6fbd857d",
   "metadata": {},
   "source": [
    "### Number of relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e8acf3fe-2a3b-4b44-97c4-9dc35098cc1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node1</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>Node2</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Assay</td>\n",
       "      <td>MEASURED_ASmMG</td>\n",
       "      <td>MGene</td>\n",
       "      <td>25878862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGene</td>\n",
       "      <td>IS_ORTHOLOG_MGiG</td>\n",
       "      <td>Gene</td>\n",
       "      <td>56712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Study</td>\n",
       "      <td>PERFORMED_SpAS</td>\n",
       "      <td>Assay</td>\n",
       "      <td>6178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Assay</td>\n",
       "      <td>INVESTIGATED_ASiA</td>\n",
       "      <td>Anatomy</td>\n",
       "      <td>6064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Assay</td>\n",
       "      <td>INVESTIGATED_ASiCT</td>\n",
       "      <td>CellType</td>\n",
       "      <td>1134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Assay</td>\n",
       "      <td>MEASURED_ASmMR</td>\n",
       "      <td>MethylationRegion</td>\n",
       "      <td>416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MGene</td>\n",
       "      <td>METHYLATED_IN_MGmMR</td>\n",
       "      <td>MethylationRegion</td>\n",
       "      <td>355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Mission</td>\n",
       "      <td>CONDUCTED_MIcS</td>\n",
       "      <td>Study</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MetaNode</td>\n",
       "      <td>MetaRelationship</td>\n",
       "      <td>MetaNode</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Node1         Relationship              Node2     Count\n",
       "0     Assay       MEASURED_ASmMG              MGene  25878862\n",
       "1     MGene     IS_ORTHOLOG_MGiG               Gene     56712\n",
       "2     Study       PERFORMED_SpAS              Assay      6178\n",
       "3     Assay    INVESTIGATED_ASiA            Anatomy      6064\n",
       "4     Assay   INVESTIGATED_ASiCT           CellType      1134\n",
       "5     Assay       MEASURED_ASmMR  MethylationRegion       416\n",
       "6     MGene  METHYLATED_IN_MGmMR  MethylationRegion       355\n",
       "7   Mission       CONDUCTED_MIcS              Study        74\n",
       "8  MetaNode     MetaRelationship           MetaNode         8"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (n1)-[r]->(n2)\n",
    "WITH TYPE(r)        AS Relationship,\n",
    "     count(r)       AS Count,\n",
    "     LABELS(n1)[0]  AS Node1,\n",
    "     LABELS(n2)[0]  AS Node2\n",
    "RETURN Node1, Relationship, Node2, Count\n",
    "ORDER BY Count DESC;\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "123e432b-4132-4526-a419-68d1a083aa1e",
   "metadata": {},
   "source": [
    "## Fulltext Search <a class=\"anchor\" id=\"Fulltext Search\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ba241e-ba6f-4dd9-b064-a7e0e6d3c79f",
   "metadata": {},
   "source": [
    "### Full text query by keyword or phrase\n",
    "A full text query returns all nodes that match the text query ([Query Syntax](https://lucene.apache.org/core/5_5_0/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#Overview)). For exact matches, enclose the phrase in double quotes, e.g., ```'\"muscle\"'```.\n",
    "\n",
    "[Learn more about full-text searches.](https://graphaware.com/neo4j/2019/01/11/neo4j-full-text-search-deep-dive.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c47f6333-feb9-4707-9e7b-4752a6577ee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "phrase = '\"eye\"'\n",
    "top_n = 5 # show top n hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f1258324-35f2-467c-a158-8120680acc7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>type</th>\n",
       "      <th>node</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OSD-162-ccb721824557b1437385bd84ede20971</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-162-ccb721824557b1437385bd84ede20971', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Ground Control'], 'material_2': 'eye', 'factors_1': ['Basal Control'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OSD-162-5a50e2cbc19079f36169127026802387</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-162-5a50e2cbc19079f36169127026802387', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Space Flight'], 'material_2': 'eye', 'factors_1': ['Basal Control'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OSD-162-b087b97d7b5d2326e98bcb570a4e75ed</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-162-b087b97d7b5d2326e98bcb570a4e75ed', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Ground Control'], 'material_2': 'eye', 'factors_1': ['Space Flight'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OSD-162-c7b1cabb29d13649b49721b6ee82da0f</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-162-c7b1cabb29d13649b49721b6ee82da0f', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Basal Control'], 'material_2': 'eye', 'factors_1': ['Space Flight'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OSD-162-66374d8e18123f6973011ef691f1051e</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-162-66374d8e18123f6973011ef691f1051e', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Space Flight'], 'material_2': 'eye', 'factors_1': ['Ground Control'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 identifier   type  \\\n",
       "0  OSD-162-ccb721824557b1437385bd84ede20971  Assay   \n",
       "1  OSD-162-5a50e2cbc19079f36169127026802387  Assay   \n",
       "2  OSD-162-b087b97d7b5d2326e98bcb570a4e75ed  Assay   \n",
       "3  OSD-162-c7b1cabb29d13649b49721b6ee82da0f  Assay   \n",
       "4  OSD-162-66374d8e18123f6973011ef691f1051e  Assay   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                               node  \n",
       "0  {'identifier': 'OSD-162-ccb721824557b1437385bd84ede20971', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Ground Control'], 'material_2': 'eye', 'factors_1': ['Basal Control'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}  \n",
       "1    {'identifier': 'OSD-162-5a50e2cbc19079f36169127026802387', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Space Flight'], 'material_2': 'eye', 'factors_1': ['Basal Control'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}  \n",
       "2   {'identifier': 'OSD-162-b087b97d7b5d2326e98bcb570a4e75ed', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Ground Control'], 'material_2': 'eye', 'factors_1': ['Space Flight'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}  \n",
       "3    {'identifier': 'OSD-162-c7b1cabb29d13649b49721b6ee82da0f', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Basal Control'], 'material_2': 'eye', 'factors_1': ['Space Flight'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}  \n",
       "4   {'identifier': 'OSD-162-66374d8e18123f6973011ef691f1051e', 'material_name_1': 'eye', 'technology': 'RNA Sequencing (RNA-Seq)', 'material_name_2': 'eye', 'measurement': 'transcription profiling', 'factors_2': ['Space Flight'], 'material_2': 'eye', 'factors_1': ['Ground Control'], 'material_1': 'eye', 'name': 'OSD-162_transcription-profiling_rna-sequencing-(rna-seq)_illumina', 'material_id_1': 'UBERON:0000970', 'material_id_2': 'UBERON:0000970'}  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "CALL db.index.fulltext.queryNodes(\"fulltext\", $phrase) YIELD node, score\n",
    "RETURN node.identifier AS identifier, LABELS(node)[0] AS type, node;\n",
    "\"\"\"\n",
    "graph.run(query, phrase=phrase).to_data_frame().head(top_n)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfc3758f-1dd9-4040-a3f3-efd795fe184e",
   "metadata": {},
   "source": [
    "### Full text query using boolean operators\n",
    "The full text query supports a variety of query types, including fuzzy, proximity, and range queries, as well as boolean operators ([Query Syntax](https://lucene.apache.org/core/5_5_0/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#Overview)). The following example uses a query with an ```AND``` operator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a1ef8c30-0136-48e8-bdd6-32c09325251f",
   "metadata": {},
   "outputs": [],
   "source": [
    "phrase = 'liver AND \"DNA methylation profiling\"'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c7e0f52d-d180-4620-8a14-4ef54c4dbecf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>type</th>\n",
       "      <th>node</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OSD-47-58ae5316b6c6291b4443efe11ddae112</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-47-58ae5316b6c6291b4443efe11ddae112', 'material_name_1': 'liver', 'technology': 'Whole Genome Bisulfite Sequencing', 'material_name_2': 'liver', 'measurement': 'DNA methylation profiling', 'factors_2': ['Ground Control'], 'material_2': 'Liver', 'factors_1': ['Basal Control'], 'material_1': 'Liver', 'name': 'OSD-47_dna-methylation-profiling_whole-genome-bisulfite-sequencing_illumina', 'material_id_1': 'UBERON:0002107', 'material_id_2': 'UBERON:0002107'}</td>\n",
       "      <td>9.641196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OSD-47-d2d25640688118e0d80f292eb04ba995</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-47-d2d25640688118e0d80f292eb04ba995', 'material_name_1': 'liver', 'technology': 'Whole Genome Bisulfite Sequencing', 'material_name_2': 'liver', 'measurement': 'DNA methylation profiling', 'factors_2': ['Space Flight'], 'material_2': 'Liver', 'factors_1': ['Ground Control'], 'material_1': 'Liver', 'name': 'OSD-47_dna-methylation-profiling_whole-genome-bisulfite-sequencing_illumina', 'material_id_1': 'UBERON:0002107', 'material_id_2': 'UBERON:0002107'}</td>\n",
       "      <td>9.641196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OSD-47-06f82ef81fe1f5ebba56658f21d898cf</td>\n",
       "      <td>Assay</td>\n",
       "      <td>{'identifier': 'OSD-47-06f82ef81fe1f5ebba56658f21d898cf', 'material_name_1': 'liver', 'technology': 'Whole Genome Bisulfite Sequencing', 'material_name_2': 'liver', 'measurement': 'DNA methylation profiling', 'factors_2': ['Space Flight'], 'material_2': 'Liver', 'factors_1': ['Basal Control'], 'material_1': 'Liver', 'name': 'OSD-47_dna-methylation-profiling_whole-genome-bisulfite-sequencing_illumina', 'material_id_1': 'UBERON:0002107', 'material_id_2': 'UBERON:0002107'}</td>\n",
       "      <td>9.641196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                identifier   type  \\\n",
       "0  OSD-47-58ae5316b6c6291b4443efe11ddae112  Assay   \n",
       "1  OSD-47-d2d25640688118e0d80f292eb04ba995  Assay   \n",
       "2  OSD-47-06f82ef81fe1f5ebba56658f21d898cf  Assay   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           node  \\\n",
       "0  {'identifier': 'OSD-47-58ae5316b6c6291b4443efe11ddae112', 'material_name_1': 'liver', 'technology': 'Whole Genome Bisulfite Sequencing', 'material_name_2': 'liver', 'measurement': 'DNA methylation profiling', 'factors_2': ['Ground Control'], 'material_2': 'Liver', 'factors_1': ['Basal Control'], 'material_1': 'Liver', 'name': 'OSD-47_dna-methylation-profiling_whole-genome-bisulfite-sequencing_illumina', 'material_id_1': 'UBERON:0002107', 'material_id_2': 'UBERON:0002107'}   \n",
       "1   {'identifier': 'OSD-47-d2d25640688118e0d80f292eb04ba995', 'material_name_1': 'liver', 'technology': 'Whole Genome Bisulfite Sequencing', 'material_name_2': 'liver', 'measurement': 'DNA methylation profiling', 'factors_2': ['Space Flight'], 'material_2': 'Liver', 'factors_1': ['Ground Control'], 'material_1': 'Liver', 'name': 'OSD-47_dna-methylation-profiling_whole-genome-bisulfite-sequencing_illumina', 'material_id_1': 'UBERON:0002107', 'material_id_2': 'UBERON:0002107'}   \n",
       "2    {'identifier': 'OSD-47-06f82ef81fe1f5ebba56658f21d898cf', 'material_name_1': 'liver', 'technology': 'Whole Genome Bisulfite Sequencing', 'material_name_2': 'liver', 'measurement': 'DNA methylation profiling', 'factors_2': ['Space Flight'], 'material_2': 'Liver', 'factors_1': ['Basal Control'], 'material_1': 'Liver', 'name': 'OSD-47_dna-methylation-profiling_whole-genome-bisulfite-sequencing_illumina', 'material_id_1': 'UBERON:0002107', 'material_id_2': 'UBERON:0002107'}   \n",
       "\n",
       "      score  \n",
       "0  9.641196  \n",
       "1  9.641196  \n",
       "2  9.641196  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "CALL db.index.fulltext.queryNodes(\"fulltext\", $phrase) YIELD node, score\n",
    "RETURN node.identifier AS identifier, LABELS(node)[0] AS type, node, score\n",
    "ORDER BY type;\n",
    "\"\"\"\n",
    "graph.run(query, phrase=phrase).to_data_frame()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
